src/dvclive/monitor_system.py

```diff
@@ -0,0 +1,158 @@
+import logging
+from typing import Dict, Union, Optional, List, Tuple
+from pathlib import Path
+
+import psutil
+from statistics import mean
+from threading import Event, Thread
+from funcy import merge_with
+
+from .error import InvalidDataTypeError
+
+
+logger = logging.getLogger("dvclive")
+MEGABYTES_DIVIDER = 1024.0**2
+GIGABYTES_DIVIDER = 1024.0**3
+
+MINIMUM_CPU_USAGE_TO_BE_ACTIVE = 20
+
+
+class _MonitorSystem:
+    """
+    Monitor system resources and log them to DVC Live.
+    Use a separate thread to call a `_get_metrics` function at  fix interval and
+    aggregate the results of this sampling using the average.
+    """
+
+    _plot_blacklist_prefix: Tuple = ()
+
+    def __init__(
+        self,
+        interval: float = 0.5,
+        num_samples: int = 10,
+        plot: bool = True,
+    ):
+        if not isinstance(interval, (int, float)):
+            raise InvalidDataTypeError("MonitorSystem.interval", type(interval))
+        if not isinstance(num_samples, int):
+            raise InvalidDataTypeError("MonitorSystem.sample", type(num_samples))
+        if not isinstance(plot, bool):
+            raise InvalidDataTypeError("MonitorSystem.plot", type(plot))
+
+        self._interval = interval  # seconds
+        self._nb_samples = num_samples
+        self._plot = plot
+        self._warn_user = True
+
+    def __call__(self, live):
+        self._live = live
+        self._shutdown_event = Event()
+        Thread(
+            target=self._monitoring_loop,
+        ).start()
+
+    def _monitoring_loop(self):
+        while not self._shutdown_event.is_set():
+            self._metrics = {}
+            for _ in range(self._nb_samples):
+                last_metrics = {}
+                try:
+                    last_metrics = self._get_metrics()
+                except psutil.Error:
+                    if self._warn_user:
+                        logger.exception("Failed to monitor CPU metrics")
+                        self._warn_user = False
+
+                self._metrics = merge_with(sum, self._metrics, last_metrics)
+                self._shutdown_event.wait(self._interval)
+                if self._shutdown_event.is_set():
+                    break
+            for name, values in self._metrics.items():
+                blacklisted = any(
+                    name.startswith(prefix) for prefix in self._plot_blacklist_prefix
+                )
+                self._live.log_metric(
+                    name,
+                    values / self._nb_samples,
+                    timestamp=True,
+                    plot=None if blacklisted else self._plot,
+                )
+
+    def _get_metrics(self):
+        pass
+
+    def end(self):
+        self._shutdown_event.set()
+
+
+class MonitorCPU(_MonitorSystem):
+    _plot_blacklist_prefix: Tuple = (
+        "system/cpu/count",
+        "system/ram/total (GB)",
```

Comment:
we need to make all this names as const, or even introduce a scheme (voluptuous, etc) ...since we are using these in a few places

Reply:
I used `voluptuous` to make the test more readable. But I don't see how you want to use it in this file. Do you want to check the schema every time we gather data from psutil? 

I'll replace the string with constant, because I see the direct benefits of it. But I would love to know how you wanted to use voluptuous in that usecase :)